src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -200,41 +234,96 @@ uct_cuda_ipc_open_memhandle_vmm(uct_cuda_ipc_rkey_t *key,
                 CU_MEM_HANDLE_TYPE_FABRIC, 0));
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(&dptr,
-                key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(mapped_addr,
+                *mpool, (CUmemPoolPtrExportData*)&key->ph.ptr));
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
     status = uct_cuda_ipc_init_access_desc(&access_desc);
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolSetAccess(key->ph.pool,
-                &access_desc, 1));
+    status = UCT_CUDADRV_FUNC_LOG_ERR(
+                cuMemPoolSetAccess(*mpool, &access_desc, 1));
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
-    *mapped_addr = dptr;
-
     return UCS_OK;
+
+err:
+    return status;
+}
+
+static ucs_status_t
+uct_cuda_ipc_open_memhandle_mempool(uct_cuda_ipc_rkey_t *key,
+                                    CUdeviceptr *mapped_addr)
+{
+    khash_t(cuda_ipc_rem_mpool_cache) *hash = &uct_cuda_ipc_rem_mpool_cache.hash;
+    CUmemFabricHandle hkey                  = key->ph.handle.fabric_handle;
+    ucs_status_t status                     = UCS_OK;
+    CUmemoryPool mpool;
+    khiter_t khiter;
+    int khret;
+
+    pthread_rwlock_rdlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+    khiter = kh_get(cuda_ipc_rem_mpool_cache, hash, hkey);
+    if (ucs_likely(khiter != kh_end(hash))) {
+        key->ph.pool = kh_val(hash, khiter);
+        ucs_assert(key->ph.pool != NULL);
+
+        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(mapped_addr,
+                    key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
+        goto out;
+    }
+
+    /* acquire writer lock as we did not find cached mpool in hashmap */
+    pthread_rwlock_unlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+    pthread_rwlock_wrlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+
+    khiter = kh_put(cuda_ipc_rem_mpool_cache, hash, hkey, &khret);
+    if ((khret == UCS_KH_PUT_BUCKET_EMPTY) ||
+        (khret == UCS_KH_PUT_BUCKET_CLEAR)) {
+        status = cuda_ipc_rem_mpool_cache_create(key, &mpool, mapped_addr);
+        if (status != UCS_OK) {
+            /* Remove stale key from the hash if failed to construct new value.
+             * Otherwise next lookup will return NULL value. */
+            kh_del(cuda_ipc_rem_mpool_cache, hash, khiter);
+            goto out;
+        }
+
+        kh_val(hash, khiter) = mpool;
+        key->ph.pool         = mpool;
+    } else if (khret == UCS_KH_PUT_KEY_PRESENT) {
+        key->ph.pool = kh_val(hash, khiter);
+        status       = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(
+                            mapped_addr, key->ph.pool,
+                            (CUmemPoolPtrExportData*)&key->ph.ptr));
```

Comment:
seems cuMemPoolImportPointer is called in 3 different flows:
1. value is found using kh_get
2. value is added in cuda_ipc_rem_mpool_cache_create
3. value is found using kh_put (here)

can we unite these flows, so it will be called once, after all the khash checks? for example add label "out_import_pointer:"

Reply:
@yosefe This is required order as it stands. Can we keep this flow for now and we can reconsider post-release?